lib/spack/spack/cmd/__init__.py

```diff
@@ -177,6 +178,28 @@ def parse_specs(
     if not concretize:
         return specs
 
+    # Special case for concretizing a single spec
+    if len(specs) == 1:
+        return [specs[0].concretized()]
+
+    # Special case if every spec has an abstract spec
+    if all(spec.concrete or spec.abstract_hash for spec in specs):
+        ret = [s.concretized() for s in specs]
+
+        unify = spack.config.get("concretizer:unify", False)
+        if unify is True:  # True, "when_possible", False are possible values
+            specs_per_name = Counter(
+                spec.name
+                for spec in traverse.traverse_nodes(ret, deptype=("link", "run"))
```

Comment:
This is traversing by `id`, which fails because the concretized specs contain specs with the same DAG hash that do not have the same `id`